src/coreclr/jit/importer.cpp

```diff
@@ -1548,31 +1548,39 @@ GenTree* Compiler::impMethodPointer(CORINFO_RESOLVED_TOKEN* pResolvedToken, CORI
 
 GenTree* Compiler::getRuntimeContextTree(CORINFO_RUNTIME_LOOKUP_KIND kind)
 {
-    GenTree* ctxTree = nullptr;
+    GenTree* ctxTree;
 
     // Collectible types requires that for shared generic code, if we use the generic context parameter
     // that we report it. (This is a conservative approach, we could detect some cases particularly when the
     // context parameter is this that we don't need the eager reporting logic.)
     lvaGenericsContextInUse = true;
 
-    Compiler* pRoot = impInlineRoot();
-
     if (kind == CORINFO_LOOKUP_THISOBJ)
     {
         // this Object
-        ctxTree = gtNewLclvNode(pRoot->info.compThisArg, TYP_REF);
+        ctxTree = gtNewLclvNode(impInlineRoot()->info.compThisArg, TYP_REF);
         ctxTree->gtFlags |= GTF_VAR_CONTEXT;
 
         // context is the method table pointer of the this object
         ctxTree = gtNewMethodTableLookup(ctxTree);
     }
     else
     {
-        assert(kind == CORINFO_LOOKUP_METHODPARAM || kind == CORINFO_LOOKUP_CLASSPARAM);
+        assert((kind == CORINFO_LOOKUP_METHODPARAM) || (kind == CORINFO_LOOKUP_CLASSPARAM));
 
-        // Exact method descriptor as passed in
-        ctxTree = gtNewLclvNode(pRoot->info.compTypeCtxtArg, TYP_I_IMPL);
-        ctxTree->gtFlags |= GTF_VAR_CONTEXT;
+        if (compIsForInlining() && (impInlineInfo->iciCallInstParam != nullptr))
+        {
+            // Grab the generic context from the callsite for current inlinee.
+            // For runtime lookups, the generic context is always spilled to a local
+            assert(impInlineInfo->iciCallInstParam->OperIs(GT_LCL_VAR));
```

Comment:
IMO, it would be cleaner to do this assert once and store the local number instead of the `GenTree*`.

Reply:
I don't think I can assert untill I need an actual runtime lookup. the context may still be GT_CNS_INT/IND for other cases. the current shape allows me better validation